src/tactic/compute_degree.lean

```diff
@@ -3,33 +3,206 @@ Copyright (c) 2022 Damiano Testa. All rights reserved.
+by rwa nat_degree_add_eq_left_of_nat_degree_lt (dg.trans_lt (nat.lt_of_succ_le df.ge))
+
+lemma nat_degree_add_right_succ (n : ℕ) (f g : polynomial R)
+  (df : f.nat_degree ≤ n) (dg : g.nat_degree = n.succ) :
+  (f + g).nat_degree = n.succ :=
+by rwa nat_degree_add_eq_right_of_nat_degree_lt (df.trans_lt (nat.lt_of_succ_le dg.ge))
+
+lemma nat_degree_eq_of_le_of_coeff_ne_zero {n : ℕ} {f : polynomial R}
+  (fn : f.nat_degree ≤ n) (f0 : f.coeff n ≠ 0) :
+  f.nat_degree = n :=
+le_antisymm fn (le_nat_degree_of_ne_zero f0)
+
+lemma monic_of_le_of_coeff_eq_one [nontrivial R] {n : ℕ} {f : polynomial R}
+  (fn : f.nat_degree ≤ n) (f0 : f.coeff n = 1) :
+  f.monic :=
+begin
+  rw [monic, ← f0, leading_coeff],
+  congr,
+  refine nat_degree_eq_of_le_of_coeff_ne_zero fn (λ h, (one_ne_zero (f0.symm.trans h))),
+end
+
+lemma coeff_C_eq_zero_of_succ (n : ℕ) (a : R) :
+  (C a).coeff (n + 1) = 0 :=
+by simp [coeff_C]
+
+lemma nat_degree_eq_iff_le_and_coeff_ne_zero {n : ℕ} (n0 : n ≠ 0) (f : polynomial R) :
+  f.nat_degree = n ↔ f.nat_degree ≤ n ∧ f.coeff n ≠ 0 :=
+begin
+  refine ⟨_, λ ⟨fn, f0⟩, nat_degree_eq_of_le_of_coeff_ne_zero fn f0⟩,
+  rintro rfl,
+  exact ⟨rfl.le, leading_coeff_ne_zero.mpr (ne_zero_of_nat_degree_gt (nat.pos_of_ne_zero n0))⟩,
+end
+
+end polynomial
 
 namespace tactic
 namespace compute_degree
 open expr polynomial
 
+/--  `eval_guessing n e` takes a natural number `n` and an expression `e` and gives an
+estimate for the evaluation of `eval_expr' ℕ e`.  It is tailor made for estimating degrees of
+polynomials.
+
+It decomposes `e` recursively as a sequence of additions, multiplications and `max`.
+On the atoms of the process, `eval_guessing` tries to use `eval_expr' ℕ`, resorting to using
+`n` if `eval_expr' ℕ` fails.
+
+For use with degree of polynomials, we mostly use `n = 0`. -/
+meta def eval_guessing (n : ℕ) : expr → tactic ℕ
+| `(%%a + %%b)   := (+) <$> eval_guessing a <*> eval_guessing b
+| `(%%a * %%b)   := (*) <$> eval_guessing a <*> eval_guessing b
+| `(max %%a %%b) := max <$> eval_guessing a <*> eval_guessing b
+| e              := eval_expr' ℕ e <|> pure n
+
+/--  If an expression `e` is an iterated suquence of `bit0` and `bit1` starting from `0` or `1`,
+then `num_to_nat e` returns `some n`, where `n` is the natural number obtained from the same
+sequence of `bit0` and `bit1` applied to `0` or `1`.  Otherwise, `num_to_nat e = none`. -/
+-- for the application, the line `| `(has_zero.zero) := some 0` step is unnecessary: the standard
+-- assumption is that the coefficient of the term of highest-looking term is non-zero.
+meta def num_to_nat : expr → option ℕ
+| `(has_zero.zero) := some 0
+| `(has_one.one)   := some 1
+| `(bit0 %%a)      := bit0 <$> num_to_nat a
+| `(bit1 %%a)      := bit1 <$> num_to_nat a
+| `(%%a + %%b)     := (+) <$> num_to_nat a <*> num_to_nat b
+| `(%%a * %%b)     := (*) <$> num_to_nat a <*> num_to_nat b
+| `(%%a ^ %%b)     := (^) <$> num_to_nat a <*> num_to_nat b
+| _ := none
```

Comment:
This feels like it must overlap heavily with norm_num.  Perhaps @digama0 can help work out precisely how?